File: workspaces/ws_005_pomodorofocus/ios/PomodoroFocusTimer/Models/SoundPlayer.swift
//
//  SoundPlayer.swift
//  PomodoroFocusTimer
//
//  Optional focus sounds (tick / white noise / rain) via AVFoundation.
//
//  NOTE: audio asset files are not bundled in this code-only deliverable.
//  Drop `tick.caf`, `whitenoise.caf` and `rain.caf` into the app target and
//  this player loops them seamlessly. With no asset present the player
//  degrades gracefully to silent mode.
//

import AVFoundation

/// Plays a looping focus sound while a phase runs.
final class SoundPlayer {

    private var player: AVAudioPlayer?
    private var current: FocusSound = .silent

    /// File name (without extension) for each non-silent sound.
    private func assetName(for sound: FocusSound) -> String? {
        switch sound {
        case .silent:     return nil
        case .tick:       return "tick"
        case .whiteNoise: return "whitenoise"
        case .rain:       return "rain"
        }
    }

    /// Begins (or switches) playback for the given sound.
    func play(_ sound: FocusSound) {
        current = sound
        guard let name = assetName(for: sound) else {
            stop()
            return
        }
        // Already playing the requested sound — nothing to do.
        if player?.isPlaying == true, player?.url?.deletingPathExtension()
            .lastPathComponent == name {
            return
        }
        stop()
        guard let url = Self.locateAsset(named: name) else { return }
        configureSession()
        do {
            let p = try AVAudioPlayer(contentsOf: url)
            p.numberOfLoops = -1   // loop forever
            p.volume = 0.6
            p.prepareToPlay()
            p.play()
            player = p
        } catch {
            player = nil
        }
    }

    /// Resumes if a non-silent sound is selected (used on foreground return).
    func resumeIfNeeded(_ sound: FocusSound) {
        if current != .silent, player?.isPlaying == false {
            player?.play()
        } else if current == .silent {
            // selection changed to silent while away
        }
    }

    func pause() { player?.pause() }

    func stop() {
        player?.stop()
        player = nil
    }

    private func configureSession() {
        let session = AVAudioSession.sharedInstance()
        // .playback so loops continue while the screen locks / app backgrounds.
        try? session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
        try? session.setActive(true)
    }

    /// Finds an audio asset, trying common loopable extensions.
    private static func locateAsset(named name: String) -> URL? {
        for ext in ["caf", "m4a", "mp3", "wav"] {
            if let url = Bundle.main.url(forResource: name, withExtension: ext) {
                return url
            }
        }
        return nil
    }
}

File: workspaces/ws_005_pomodorofocus/android/app/src/main/java/com/appfactory/pomodorofocus/audio/SoundPlayer.kt
package com.appfactory.pomodorofocus.audio

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaPlayer
import com.appfactory.pomodorofocus.model.FocusSound

/**
 * Plays an optional looping focus sound while a phase runs.
 *
 * NOTE: audio assets are not bundled in this code-only deliverable. Drop
 * `tick.ogg`, `whitenoise.ogg` and `rain.ogg` into `app/src/main/res/raw/`
 * and this player loops them. With no asset present it degrades gracefully
 * to silent mode.
 */
class SoundPlayer(private val context: Context) {

    private var player: MediaPlayer? = null
    private var current: FocusSound = FocusSound.SILENT

    /** Resource name (without extension) expected in res/raw for each sound. */
    private fun rawName(sound: FocusSound): String? = when (sound) {
        FocusSound.SILENT -> null
        FocusSound.TICK -> "tick"
        FocusSound.WHITE_NOISE -> "whitenoise"
        FocusSound.RAIN -> "rain"
    }

    /** Starts (or switches to) the given looping sound. */
    fun play(sound: FocusSound) {
        if (sound == current && player?.isPlaying == true) return
        current = sound
        stop()
        val name = rawName(sound) ?: return
        val resId = context.resources.getIdentifier(name, "raw", context.packageName)
        if (resId == 0) return // asset not bundled — stay silent
        player = MediaPlayer.create(context, resId)?.apply {
            isLooping = true
            setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
                    .build()
            )
            setVolume(0.6f, 0.6f)
            start()
        }
    }

    /** Pauses playback, retaining position (used when the timer pauses). */
    fun pause() {
        player?.takeIf { it.isPlaying }?.pause()
    }

    /** Resumes a previously selected non-silent sound. */
    fun resume() {
        if (current != FocusSound.SILENT) {
            player?.takeIf { !it.isPlaying }?.start()
        }
    }

    /** Stops and releases the player. */
    fun stop() {
        player?.run {
            if (isPlaying) stop()
            release()
        }
        player = null
    }
}
